community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/pipes/FullPruningVarLengthExpandPipe.scala

```diff
@@ -0,0 +1,395 @@
+
+      updatePrevFullExpandDepth()
+
+      if (!nodeState.isEmitted && pathLength >= self.min) {
+        nodeState.isEmitted = true
+        expandMap.put(node.getId, nodeState)
+        node
+      } else {
+        expandMap.put(node.getId, nodeState)
+        null
+      }
+    }
+
+    private def hasRelationship = relationshipCursor < nodeState.rels.length
+
+    private def nextRelationship() = {
+      val next = relationshipCursor
+      relationshipCursor += 1
+      next
+    }
+
+    private def haveFullyExploredTheRemainingStepsBefore(i: Int) = {
+      val stepsLeft = self.max - pathLength
+      nodeState.depths(i) >= stepsLeft
+    }
+
+    def updatePrevFullExpandDepth() = {
+      if ( pathLength > 0 ) {
+        val requiredStepsFromPrev = math.max(0, self.min - pathLength + 1)
+        if (requiredStepsFromPrev <= 1 || nodeState.isEmitted) {
+          prevNodeState.updateFullExpandDepth(prevLocalRelIndex, currentOutgoingFullExpandDepth() + 1)
+        }
+      }
+    }
+
+    private def currentOutgoingFullExpandDepth(): Int = {
+      assert(pathLength > 0)
+      if (pathLength == self.max) 0
+      else {
+        // The maximum amount of steps that have been fully explored is the minimum full expand depth of the
+        // outgoing relationships. The incoming relationship is not included, as that is the relationship that will be
+        // updated.
+        val incomingRelationship = path(pathLength - 1)
+        nodeState.minOutgoingDepth(incomingRelationship)
+      }
+    }
+
+    private def initiate() = {
+      nodeState = expandMap.get(node.getId)
+      if (nodeState == UNINITIALIZED) {
+        nodeState = new NodeState()
+      }
+    }
+  }
+
+  trait CheckPath {
+    def pathLength: Int
+
+    val path: Array[Long]
+
+    def seenRelationshipInPath(r: Long): Boolean = {
+      if (pathLength == 0) return false
+      var idx = 0
+      while (idx < pathLength) {
+        if (path(idx) == r) return true
+        idx += 1
+      }
+      false
+    }
+  }
+
+  object NodeState {
+    val UNINITIALIZED: NodeState = null
+
+    val NOOP = new NodeState() {
+      override def minOutgoingDepth(incomingRelId: Long): Int = 0
+      override def updateFullExpandDepth(relIndex:Int, depth:Int ) = {}
+    }
+  }
+
+  /**
+    * The state of expansion for one node
+    */
+  class NodeState() {
+    // All relationships that connect to this node, filtered by the var-length predicates
+    var rels: Array[Relationship] = _
+    // The fully expanded depth for each relationship in rels
+    var depths:Array[Byte] = _
+    // True if this node has been emitted before
+    var isEmitted = false
+
+    /**
+      * Computes the minimum outgoing full expanded depth.
+      *
+      * @param incomingRelId id of the incoming relationship, which is omitted from this computation
+      * @return the full expand depth
+      */
+    def minOutgoingDepth(incomingRelId: Long): Int = {
+      var max = Constants.VERY_BIG_VALUE
```

Comment:
How about using `Integer.MAX_VALUE` instead of this arbitrarily chosen `VERY_BIG_VALUE`?

This is the only place `VERY_BIG_VALUE` is used, so if that is changed, we no longer need to declare that constant - making the code easier to read (because a reader no longer needs to try to understand why the value is what it is).